src/lib/communication/DY.Lib.Communication.Core.Invariants.fst

```diff
@@ -0,0 +1,188 @@
+}
+#pop-options
+
+val sign_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & sign_crypto_predicate)
+let sign_crypto_predicates_communication_layer_and_tag #cusages =
+  (comm_layer_sign_tag, sign_crypto_predicates_communication_layer)
+
+val has_communication_layer_crypto_predicates:
+  {|crypto_invariants|} ->
+  prop
+let has_communication_layer_crypto_predicates #cinvs =
+  has_pke_predicate pke_crypto_predicates_communication_layer_and_tag /\
+  has_sign_predicate sign_crypto_predicates_communication_layer_and_tag
+
+(*** Event Predicates ***)
+
+noeq
+type comm_higher_layer_event_preds = {
+  send_conf: tr:trace -> sender:principal -> receiver:principal -> payload:bytes -> prop;
+  send_conf_later: 
+    tr1:trace -> tr2:trace ->
+    sender:principal -> receiver:principal -> payload:bytes ->
+    Lemma
+    (requires
+      send_conf tr1 sender receiver payload /\
+      bytes_well_formed tr1 payload /\
+      tr1 <$ tr2
+    )
+    (ensures send_conf tr2 sender receiver payload)
+  ;
+  send_auth: tr:trace -> sender:principal -> payload:bytes -> prop;
+  send_auth_later: 
+    tr1:trace -> tr2:trace ->
+    sender:principal -> payload:bytes ->
+    Lemma
+    (requires
+      send_auth tr1 sender payload /\
+      bytes_well_formed tr1 payload /\
+      tr1 <$ tr2
+    )
+    (ensures send_auth tr2 sender payload)
+  ;
+  send_conf_auth: tr:trace -> sender:principal -> receiver:principal -> payload:bytes -> prop;
+  send_conf_auth_later: 
+    tr1:trace -> tr2:trace ->
+    sender:principal -> receiver:principal -> payload:bytes ->
+    Lemma
+    (requires
+      send_conf_auth tr1 sender receiver payload /\
+      bytes_well_formed tr1 payload /\
+      tr1 <$ tr2
+    )
+    (ensures send_conf_auth tr2 sender receiver payload)
+}
+
+val default_comm_higher_layer_event_preds: comm_higher_layer_event_preds
+let default_comm_higher_layer_event_preds = {
+  send_conf = (fun tr sender receiver payload -> True);
+  send_conf_later = (fun tr1 tr2 sender receiver payload -> ());
+  send_auth = (fun tr sender payload -> True);
+  send_auth_later = (fun tr1 tr2 sender payload -> ());
+  send_conf_auth = (fun tr sender receiver payload -> True);
+  send_conf_auth_later = (fun tr1 tr2 sender receiver payload -> ())
+}
+
+#push-options "--ifuel 1 --fuel 0"
+let event_predicate_communication_layer 
+  {|cinvs:crypto_invariants|}
+  (higher_layer_preds:comm_higher_layer_event_preds) : 
+  event_predicate communication_event =
+  fun tr prin e ->
+    (match e with
+    | CommConfSendMsg sender receiver payload -> (
+      is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload /\
+      higher_layer_preds.send_conf tr sender receiver payload
+    )
+    | CommConfReceiveMsg receiver payload -> (
+      exists sender.
+        is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload /\
```

Comment:
I think this line is unnecessary --- in the honest case, we get this from the event predicate for `CommConfSendMsg`, and in the corrupt case, it's immediate because the payload is publishable.